File: dispatch/core/RemoteFunction/FuturesTable.cc
#include "FuturesTable.hh"
#include "AbstractFuture.hh"


namespace claid {

void FuturesTable::addFuture(std::shared_ptr<AbstractFuture> future)
{
    std::unique_lock<std::mutex>(this->lock);
    futures.insert(std::make_pair(future->getUniqueIdentifier(), future));
}

bool FuturesTable::removeFuture(FutureUniqueIdentifier futureIdentifier)
{
    std::unique_lock<std::mutex>(this->lock);

    auto it = futures.find(futureIdentifier);
    if(it != futures.end())
    {
        std::shared_ptr<AbstractFuture> other = it->second;
     
        futures.erase(it);
        return true;
    }
    return false;
}

std::shared_ptr<AbstractFuture> FuturesTable::lookupFuture(FutureUniqueIdentifier uniqueIdentifier)
{
    std::shared_ptr<AbstractFuture> future = nullptr;

    std::unique_lock<std::mutex>(this->lock);

    auto it = futures.find(uniqueIdentifier);
    if(it != futures.end())
    {
        future = it->second;
    }
    return future;
}

}

File: dispatch/python/claid/remote_function/futures_table.py
import threading

class FuturesTable:
    def __init__(self):
        self.futures = {}
        self.lock = threading.Lock()

    def add_future(self, future):
        with self.lock:
            self.futures[future.get_unique_identifier().to_string()] = future

    def remove_future(self, future):
        future_identifier = future.get_unique_identifier().to_string()

        with self.lock:
            if future_identifier in self.futures:
                if self.futures[future_identifier] is future:
                    del self.futures[future_identifier]
                    return True
        return False

    def lookup_future(self, unique_identifier):
        with self.lock:
            return self.futures.get(unique_identifier.to_string(), None)
